nscplugin/src/main/scala-next/scala/scalanative/nscplugin/NirGenExpr.scala

```diff
@@ -0,0 +1,2642 @@
+
+  class ExprBuffer(using fresh: Fresh) extends FixupBuffer {
+    buf =>
+    def genExpr(tree: Tree): Val = {
+      tree match {
+        case EmptyTree      => Val.Unit
+        case ValTree(value) => value
+        case ContTree(f)    => f()
+        case tree: Apply =>
+          val updatedTree = lazyValsAdapter.transformApply(tree)
+          genApply(updatedTree)
+        case tree: Assign         => genAssign(tree)
+        case tree: Block          => genBlock(tree)
+        case tree: Closure        => genClosure(tree)
+        case tree: Labeled        => genLabelDef(tree)
+        case tree: Ident          => genIdent(tree)
+        case tree: If             => genIf(tree)
+        case tree: JavaSeqLiteral => genJavaSeqLiteral(tree)
+        case tree: Literal        => genLiteral(tree)
+        case tree: Match          => genMatch(tree)
+        case tree: Return         => genReturn(tree)
+        case tree: Select         => genSelect(tree)
+        case tree: This           => genThis(tree)
+        case tree: Try            => genTry(tree)
+        case tree: Typed          => genTyped(tree)
+        case tree: TypeApply      => genTypeApply(tree)
+        case tree: ValDef         => genValDef(tree)
+        case tree: WhileDo        => genWhileDo(tree)
+        case _ =>
+          throw FatalError(
+            s"""Unexpected tree in genExpr: `${tree}`
+             raw=$tree
+             pos=${tree.span}
+             """
+          )
+      }
+    }
+
+    def defnNir(using Context): NirDefinitions = NirDefinitions.get
+
+    lazy val SafeZoneHandle = new Property.Key[Val]
+
+    private def safeZoneHandle(tree: Tree): Val =
+      tree.getAttachment(SafeZoneHandle).getOrElse(Val.Null)
+
+    def genApply(app: Apply): Val = {
+      app match {
+        case Apply(_, List(sz, tree))
+            if app.fun.symbol == defnNir.SafeZoneCompat_withSafeZone =>
+          // For new expression with a specified safe zone, e.g. `new {sz} T(...)`,
+          // it's translated to `withSafeZone(sz, new T(...))` in TyperPhase.
+          tree match {
+            case Apply(Select(New(_), nme.CONSTRUCTOR), _) =>
+            case Apply(fun, _) if fun.symbol == defn.newArrayMethod =>
+            case _ => 
+              report.error(s"Unexpected tree in withSafeZone: `${tree}`", tree.srcPos)
+          }
+          // Extract the handle of `sz` and put it into the attachment of `new T(...)`.
+          val handle = genExpr(Select(sz, termName("handle")))
+          if tree.hasAttachment(SafeZoneHandle) then
+            report.warning(
+              s"Safe zone handle is already attached to ${tree}, which is unexpected.",
+              tree.srcPos
+            );
+          tree.putAttachment(SafeZoneHandle, handle)
+          // Translate `withSafeZone(sz, new T(...))` to `{ sz.checkOpen(); new T(...) }`.
+          val checkOpen = Apply(Select(sz, termName("checkOpen")), List())
```

Comment:
In theory, we could pass `SafeZone` instance instead of its handle to classalloc. This way we should've able to call `checkOpen` from within of actual allocation

Reply:
That sounds cool! But I'm wonder how to do so. Like, somehow call the `checkOpen` when lowering `classalloc`?

Reply:
Yes, It might look like this: 
```patch
diff --git a/tools/src/main/scala/scala/scalanative/codegen/Lower.scala b/tools/src/main/scala/scala/scalanative/codegen/Lower.scala
index 9f8d083b0..066739c11 100644
@@ -770,8 +770,32 @@ object Lower {
       assert(size == size.toInt)
 
       zoneHandle match {
-        case Some(zoneHandle) =>
-          assert(zoneHandle != Val.Null, "ZoneHandle is null")
+        case Some(instance) =>
+          assert(instance != Val.Null, "ZoneHandle is null")
+          val checkOpenMethod, handleAccessor = Val.Local(fresh(), Type.Ptr)
+          // Perform virtual dispatch of checkOpen method and call it
+          genMethodOp(
+            buf,
+            checkOpenMethod.name,
+            Op.Method(instance, SafeZoneCheckOpen.sig)
+          )
+          buf.call(SafeZoneCheckOpenSig, checkOpenMethod, Seq(instance), unwind)
+        
+          // Perform virtual dispatch of handle accessor and call it
+          // Currently it's not working correctly - optimizer might replace method dispatch op with direct fieldload becouse there is only 1 implementation of SafeZone
+          genMethodOp(
+            buf,
+            handleAccessor.name,
+            Op.Method(instance, SafeZoneHandle.sig)
+            )
+          val zoneHandle = buf.call(
+            SafeZoneHandleSig,
+            handleAccessor,
+            Seq(instance),
+            unwind
+          )
+
+          // Call actual zone allocation using resolved zoneHandle
           buf.let(
             n,
             Op.Call(
@@ -1422,6 +1446,19 @@ object Lower {
   val RuntimeNull = Type.Ref(Global.Top("scala.runtime.Null$"))
   val RuntimeNothing = Type.Ref(Global.Top("scala.runtime.Nothing$"))
 
+  val SafeZone = Type.Ref(Global.Top("scala.scalanative.SafeZone"))
+
+  val SafeZoneCheckOpenSig = Type.Function(Seq(SafeZone), Type.Unit)
+  val SafeZoneCheckOpen =
+    SafeZone.name.member(Sig.Method("checkOpen", Seq(Type.Unit)))
+
+  val SafeZoneHandleSig = Type.Function(Seq(SafeZone), Type.Ptr)
+  val SafeZoneHandle =
+    SafeZone.name.member(Sig.Method("handle", Seq(Type.Ptr)))
+
   val injects: Seq[Defn] = {
     implicit val pos = Position.NoPosition
     val buf = mutable.UnrolledBuffer.empty[Defn]
@@ -1467,6 +1504,9 @@ object Lower {
     buf += throwNoSuchMethod
     buf += RuntimeNull.name
     buf += RuntimeNothing.name
+    buf += SafeZoneCheckOpen
+    buf += SafeZoneHandle
     buf.toSeq
   }
 }

```

But it would be even better to add an additional method to `trait SafeZone` which would be responsible for all the required logic.
```scala
  private[scalanative] def allocImpl[T](size: CSize, cls: Class[T]): Ptr[T] = {
    this.checkOpen()
    CZone.alloc(handle, castObjectToRawPtr(cls), size)
  }
```
or 
```scala
  private[scalanative] def allocImpl[T](size: RawSize, cls: RawPtr): RawPtr = {
    this.checkOpen()
    CZone.alloc(handle, cls, size)
  }
```
Then we can add a Op.Method dispatch like in the snippet above to get correct implementation of allocImpl if there would more than 1 implementation or direct call to the only used implementation otherwise.

The one biggest improvement of moving generation of additional logic (e.g. checkOpen) to lowering phase, possibly combined with virtual method dispatch is the fact that we would be able to easily add changes to the logic if they would be needed. This would improve backward and forward compatibility of the produced NIR.